File: src/main/java/at/bitfire/dav4android/PropertyRegistry.java
package at.bitfire.dav4android;

import org.xmlpull.v1.XmlPullParser;

import java.util.HashMap;
import java.util.Map;
import java.util.ServiceLoader;

public class PropertyRegistry {

    protected Map<String, Map<String, PropertyFactory>> factories = new HashMap<>();

    static final PropertyRegistry DEFAULT = new PropertyRegistry();

    private PropertyRegistry() {
        for (PropertyFactory factory : ServiceLoader.load(PropertyFactory.class)) {
            Constants.log.debug("Registering DAV property factory for " + factory.getName());
            register(factory);
        }
    }


    public void register(PropertyFactory factory) {
        Property.Name name = factory.getName();
        Map<String, PropertyFactory> nsFactories = factories.get(name.namespace);
        if (nsFactories == null)
            factories.put(name.namespace, nsFactories = new HashMap<>());
        nsFactories.put(name.name, factory);
    }

    public Property create(Property.Name name, XmlPullParser parser) {
        Map<String, PropertyFactory> map = factories.get(name.namespace);
        if (map != null) {
            PropertyFactory factory = map.get(name.name);
            if (factory != null)
                return factory.create(parser);
        }
        return null;
    }

}

File: src/main/java/at/bitfire/dav4android/PropertyRegistry.kt
/*
 * Copyright © Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.dav4android

import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.util.*
import java.util.logging.Level

object PropertyRegistry {

    private val factories = mutableMapOf<Property.Name, PropertyFactory>()

    init {
        Constants.log.info("Registering DAV property factories")
        for (factory in ServiceLoader.load(PropertyFactory::class.java)) {
            Constants.log.fine("Registering ${factory::class.java.name} for ${factory.getName()}")
            register(factory)
        }
    }


    private fun register(factory: PropertyFactory) {
        factories[factory.getName()] = factory
    }

    fun create(name: Property.Name, parser: XmlPullParser) =
            try {
                factories[name]?.create(parser)
            } catch (e: XmlPullParserException) {
                Constants.log.log(Level.WARNING, "Couldn't parse $name", e)
                null
            }

}
